Update README.md (#19)
ITWDiagraph/BoxwriterInterop: README.md

[![.NET](https://github.com/ITWDiagraph/BoxwriterInterop/actions/workflows/dotnet.yml/badge.svg?branch=main)](https://github.com/ITWDiagraph/BoxwriterInterop/actions/workflows/dotnet.yml)

### Configuration
Use the Add Line command to add printer names and addresses to the PrinterConnections.json file.

# Functionality
## Locate
Making a UDP broadcast Request on port 2200 with the data

    {Locate BoxWriter}

The application will respond with

    {Locate BoxWriter, Address, Port, Machine Name}

- **Address**: IP address of the machine
- **Port**: Port it will respond on
- **Machine Name**: Machine name

## Get Tasks
Determines the available tasks that the printer can start

> Making a TCP Request on the port *default 2202* with the data

Reviewer: I would move this default port stuff to a "Making a call" section. You dont need to tell me the same port for every command.

Something like

## Making a request
General requests are made on TCP port `2202` (unless otherwise specified) in the format:

    {command, printerID, param1, param2, ...}

It will typically respond in the following format:

    {command, printerID, result1, result2, ...}

> Making a TCP Request on the port *default 2202* with the data

Reviewer: I think we can simplify to `Request`

    {Get tasks, PrinterName}

> The application will respond with

Reviewer: I think we can simplify to `Response`

Author: I think this makes it less clear to the reader for what we are doing

    
    {Get tasks, PrinterName, TaskName1, TaskName2, ...}
    
- **PrinterName**: Name of the printer that the command is targeting
- **TaskName**: Name of a task that is on the printer

> There will be no tasks returned if there are no tasks on the printer

Reviewer: ```suggestion
`{Get tasks, PrinterName}` is returned if there are no tasks on the printer
```

## Load Task
Prints a message

Making a TCP Request on the port *default 2202* with the data

    {Load task, PrinterName, TaskName}

The application will respond with

    {Load task, PrinterName, Result}

- **PrinterName**: Name of the printer that the command is targeting
- **Result**: Result of the command (1 for success, 0 for failure)

## Start Task
Resumes printing the current message

Making a TCP Request on the port *default 2202* with the data

    {Start task, PrinterName}

The application will respond with

    {Start task, PrinterName, Result}
    
- **PrinterName**: Name of the printer that the command is targeting
- **Result**: Result of the command (1 for success, 0 for failure)

## Idle Task

> This command pauses the current message

Reviewer: Missed this one: Uses old "This command..." style

Making a TCP Request on the port *default 2202* with the data

    {Idle task, PrinterName}

The application will respond with

    {Idle task, PrinterName, Result}
    
- **PrinterName**: Name of the printer that the command is targeting
- **Result**: Result of the command (1 for success, 0 for failure)

## Resume Task
Resumes the current message

Making a TCP Request on the port *default 2202* with the data

    {Resume task, PrinterName}

The application will respond with

    {Resume task, PrinterName, Result}
    
- **PrinterName**: Name of the printer that the command is targeting
- **Result**: Result of the command (1 for success, 0 for failure)

## Add Line
Saves a printer to the configuration file

Making a TCP Request on the port *default 2202* with the data

    {Add line, PrinterName, IPAddress}

The application will respond with
 
    {Add line, PrinterName, Result}
    
- **PrinterName**: Name of a printer that is configured
- **Result**: Result of the command (1 for success, 0 for failure)

## Get Lines

> Gets all the saved printers

Reviewer: ```suggestion
Returns all saved printer connections
```

Making a TCP Request on the port *default 2202* with the data

    {Get lines}

The application will respond with

    {Get lines, PrinterName1, PrinterName2, ...}
    
- **PrinterName**: Name of a printer that is saved

## Get User Elements
Gets the variable data available for the current message. It returns no data if there are no variables in the message.

Making a TCP Request on the port *default 2202* with the data

    {Get user elements, PrinterName}

The application will respond with

    {Get user elements, PrinterName, Prompt1, Data1, Prompt2, Data2, ...}
    
- **PrinterName**: Name of the printer
- **Prompt**: The variable prompt
- **Data**: The variable data

## Set User Elements

> Gets the variable data available for the current message. It returns no data if there are no variables in the message.

Reviewer: Move "It returns no data if there are no variables in the message." to the Data detail part

[![.NET](https://github.com/ITWDiagraph/BoxwriterInterop/actions/workflows/dotnet.yml/badge.svg?branch=main)](https://github.com/ITWDiagraph/BoxwriterInterop/actions/workflows/dotnet.yml)

### Configuration
Use the Add Line command to add printer names and addresses to the PrinterConnections.json file.

### Requests and Responses
Requests are made on TCP port `2202` (unless otherwise specified) in the format:

    {command, printerID, param1, param2, ...}

Responses will typically be in the following format:

    {command, printerID, result1, result2, ...}

# Functionality
## Locate

> Making a UDP broadcast Request on port 2200 with the data

Reviewer 2: The organization of this is odd.  The statement above says requests are made with TCP then the first functionality that we present is the only UDP command.  

I think we should organize the UDP command and the TCP commands separate.  The block talking about TCP commands is good I think it should just be below the UDP command which is setup by itself.

> {Get tasks, **_PrinterName_**}

The application will respond with

> > {Get tasks, **_PrinterName_**, **_TaskName1_**, **_TaskName2_**, ...}

Reviewer: We should clarify that the results will be empty if no files are present.

- **_PrinterName_**: Name of the printer that the command is targeting
- **_TaskName_**: Name of a task that is available on this printer

### LoadTask

> This command is used to print a message

Reviewer: I think the "This X is used..." is over-worded. Can be reduced to "Prints a message". Same idea with other documentation we've done, get to the point, skip any sentence fluff.

The application will respond with
> {Start task, **_PrinterName_**, **_Result_**}
- **_PrinterName_**: Name of the printer that the command is targeting
- **_Result_**: Result of the command (1 for success, 0 for failure)

### IdleTask
This command pauses the current message

Making a TCP Request on the port *default 2202* with the data
> {Idle task, **_PrinterName_**}

The application will respond with
> {Idle task, **_PrinterName_**, **_Result_**}
- **_PrinterName_**: Name of the printer that the command is targeting
- **_Result_**: Result of the command (1 for success, 0 for failure)

### ResumeTask
This command resumes the current message

Making a TCP Request on the port *default 2202* with the data
> {Resume task, **_PrinterName_**}

The application will respond with
> {Resume task, **_PrinterName_**, **_Result_**}
- **_PrinterName_**: Name of the printer that the command is targeting
- **_Result_**: Result of the command (1 for success, 0 for failure)

### AddLine
This command saves a printer to the configuration file

Making a TCP Request on the port *default 2202* with the data
> {Add line, **_PrinterName_**, **IPAddress**}

The application will respond with

> > {Get lines, **_PrinterName1_**, **_PrinterName2_**, ...}

Reviewer 2: This is not correct this should respond with Add line not Get lines

- **_PrinterName_**: Name of a printer that is configured

### GetLines
This request gets all the saved printers

Making a TCP Request on the port *default 2202* with the data
> {Get lines}

The application will respond with
> {Get lines, **_PrinterName1_**, **_PrinterName2_**, ...}
- **_PrinterName_**: Name of a printer that is saved

### GetUserElements
This request gets the variable data available for the current message. It returns no data if there are no variables in the message.

Making a TCP Request on the port *default 2202* with the data
> {Get user elements, **_PrinterName_**}

The application will respond with
> {Get user elements, **_PrinterName_**, **_Prompt1_**, **_Data1_**, **_Prompt2_**, **_Data2_**...}

> - **_PrinterName_**: Name of a printer that is saved

Reviewer 2: This should be
` **_PrinterName_**: Name of the printer`

- **_Prompt_**: The variable prompt
- **_Data_**: The variable data

### SetUserElements
This request gets the variable data available for the current message. It returns no data if there are no variables in the message.

Making a TCP Request on the port *default 2202* with the data
> {Set user elements, **_PrinterName_**, **_Prompt1_**, **_Data1_**, **_Prompt2_**, **_Data2_**...}

The application will respond with
> {Set user elements, **_PrinterName_**, **_Result_**}

> - **_PrinterName_**: Name of a printer that is saved

Reviewer 2: Same as above since we are not saving the printer name here.
